hsds/domain_crawl.py

```diff
@@ -0,0 +1,275 @@
+    ):
+        log.info(f"DomainCrawler.__init__  root_id: {len(objs)} objs")
+        self._app = app
+        self._action = action
+        self._max_objects_limit = max_objects_limit
+        self._params = params
+        self._max_tasks = max_tasks
+        self._q = asyncio.Queue()
+        self._obj_dict = {}
+        self.seen_ids = set()
+        if not objs:
+            log.error("no objs for crawler to crawl!")
+            raise ValueError()
+
+        for obj_id in objs:
+            log.debug(f"adding {obj_id} to the queue")
+            self._q.put_nowait(obj_id)
+        if isinstance(objs, dict):
+            self._objs = objs
+        else:
+            self._objs = None
+
+
+    async def get_attributes(self, obj_id, attr_names):
+        # get the given attributes for the obj_id
+        log.debug(f"get_attributes for {obj_id}, {len(attr_names)} attributes")
+        req = getDataNodeUrl(self._app, obj_id)
+        collection = getCollectionForId(obj_id)
+        req += f"/{collection}/{obj_id}/attributes"
+        log.debug(f"POST Attributes: {req}")
+        params = {}
+        for key in ("IncludeData", "ignore_nan", "bucket"):
+            if key in self._params:
+                value = self._params[key]
+                if isinstance(value, bool):
+                    if value:
+                        params[key] = 1  # http doesn't like booleans
+                else:
+                    params[key] = self._params[key]
+        data = {"attributes": attr_names}
+        log.debug(f"using params: {params}")
+
+        dn_json = None
+        status = 200
+        try:
+            dn_json = await http_post(self._app, req, data=data, params=params)
+        except HTTPBadRequest:
+            status = 400
+        except HTTPNotFound:
+            status = 404
+        except HTTPGone:
+            status = 410
+        except HTTPServiceUnavailable:
+            status = 503
+        except HTTPInternalServerError:
+            status = 500
+        except Exception as e:
+            log.error(f"unexpected exception from post request: {e}")
+            
+        if status == 200:
+            log.debug(f"got attributes json from dn for obj_id: {dn_json}")
+            if "attributes" not in dn_json:
+                log.error(f"DomainCrawler - expected attributes in json, but got: {dn_json}")
+                return
+            attributes = dn_json["attributes"]
+
+            if len(attributes) < len(attr_names):
+                msg = f"POST attributes requested {len(attr_names)}, "
+                msg += f"but only {len(attributes)} were returned"
+                log.warn(msg)
+            self._obj_dict[obj_id] = attributes
+        else:
+            log.warn(f"Domain crawler - got {status} status for obj_id {obj_id}")
+            self._obj_dict[obj_id] = {"status": status}
+       
+
+    async def put_attributes(self, obj_id, attr_items):
+        # get the given attributes for the obj_id
```

Comment:
`get the given attributes for the obj_id` -> `write the given attributes to the obj_id`

Reply:
fixed